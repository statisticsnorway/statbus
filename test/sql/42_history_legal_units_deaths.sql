SET datestyle TO 'ISO, DMY';

BEGIN;

\echo "Setting up Statbus to test enterprise grouping and primary"

-- A Super User configures statbus.
CALL test.set_user_from_email('test.super@statbus.org');

\echo "User selected the Activity Category Standard"
INSERT INTO settings(activity_category_standard_id,only_one_setting)
SELECT id, true FROM activity_category_standard WHERE code = 'nace_v2.1'
ON CONFLICT (only_one_setting)
DO UPDATE SET
   activity_category_standard_id =(SELECT id FROM activity_category_standard WHERE code = 'nace_v2.1')
   WHERE settings.id = EXCLUDED.id;
;
SELECT acs.code
  FROM public.settings AS s
  JOIN activity_category_standard AS acs
    ON s.activity_category_standard_id = acs.id;

\echo "User uploads the sample activity categories"
\copy public.activity_category_available_custom(path,name,description) FROM 'samples/norway/activity_category/activity_category_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.activity_category_available;

\echo "User uploads the sample regions"
\copy public.region_upload(path, name) FROM 'samples/norway/regions/norway-regions-2024.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.region;

\echo "User uploads the sample legal forms"
\copy public.legal_form_custom_only(code,name) FROM 'samples/norway/legal_form/legal_form_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.legal_form_available;

\echo "User uploads the sample sectors"
\copy public.sector_custom_only(path,name,description) FROM 'samples/norway/sector/sector_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.sector_available;

SAVEPOINT before_loading_units;

\echo "Test deaths at the end of the year"


SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;

\echo "User uploads the legal units"
\copy public.import_legal_unit_era(valid_from,valid_to,tax_ident,name,birth_date,death_date,physical_address_part1,physical_postcode,physical_postplace,physical_region_code,physical_country_iso_2,postal_address_part1,postal_postcode,postal_postplace,postal_region_code,postal_country_iso_2,primary_activity_category_code,secondary_activity_category_code,sector_code,legal_form_code) FROM 'test/data/42_legal-units-deaths-end-of-year.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);


SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;

\echo "Refreshing materialized views"
-- Exclude the refresh_time_ms as it will vary.
SELECT view_name FROM statistical_unit_refresh_now();

\echo "Check statistical units"
SELECT external_idents ->> 'tax_ident' as tax_ident, name, valid_after, valid_from, valid_to, birth_date, death_date
FROM public.statistical_unit
WHERE unit_type = 'legal_unit';

\echo "Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year'
AND year < 2013
AND unit_type = 'legal_unit';


\echo "Check statistical unit history by year-month - deaths should be 1 for year-month 2011-12 and 2012-12"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year-month'
AND year < 2013
AND unit_type = 'legal_unit';


\x


ROLLBACK TO before_loading_units;

\echo "Test deaths at the end of the first month"

\x
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;

\echo "User uploads the legal units"
\copy public.import_legal_unit_era(valid_from,valid_to,tax_ident,name,birth_date,death_date,physical_address_part1,physical_postcode,physical_postplace,physical_region_code,physical_country_iso_2,postal_address_part1,postal_postcode,postal_postplace,postal_region_code,postal_country_iso_2,primary_activity_category_code,secondary_activity_category_code,sector_code,legal_form_code) FROM 'test/data/42_legal-units-deaths-end-of-first-month.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);

SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;

\echo "Refreshing materialized views"
-- Exclude the refresh_time_ms as it will vary.
SELECT view_name FROM statistical_unit_refresh_now();

\echo "Check statistical units"
SELECT external_idents ->> 'tax_ident' as tax_ident, name, valid_after, valid_from, valid_to, birth_date, death_date
FROM public.statistical_unit
WHERE unit_type = 'legal_unit';

\echo "Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year'
AND year < 2013
AND unit_type = 'legal_unit';

\echo "Check statistical unit history by year-month - deaths should be 1 for year-month 2011-1 and 2012-12"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year-month'
AND year < 2013
AND unit_type = 'legal_unit';


\x


ROLLBACK TO before_loading_units;

\echo "Test deaths in the start of a month"

\x
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;

\echo "User uploads the legal units"
\copy public.import_legal_unit_era(valid_from,valid_to,tax_ident,name,birth_date,death_date,physical_address_part1,physical_postcode,physical_postplace,physical_region_code,physical_country_iso_2,postal_address_part1,postal_postcode,postal_postplace,postal_region_code,postal_country_iso_2,primary_activity_category_code,secondary_activity_category_code,sector_code,legal_form_code) FROM 'test/data/42_legal-units-deaths-start-of-month.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);

SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;

\echo "Refreshing materialized views"
-- Exclude the refresh_time_ms as it will vary.
SELECT view_name FROM statistical_unit_refresh_now();

\echo "Check statistical units"
SELECT external_idents ->> 'tax_ident' as tax_ident, name, valid_after, valid_from, valid_to, birth_date, death_date
FROM public.statistical_unit
WHERE unit_type = 'legal_unit';

\echo "Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year'
AND year < 2013
AND unit_type = 'legal_unit';

\echo "Check statistical unit history by year-month - deaths should be 1 for year-month 2011-1 and 2012-1"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year-month'
AND year < 2013
AND unit_type = 'legal_unit';



\x

ROLLBACK;
