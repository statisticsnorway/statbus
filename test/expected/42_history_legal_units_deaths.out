SET datestyle TO 'ISO, DMY';
BEGIN;
\i test/setup.sql
-- While the datestyle is set for the database, the pg_regress tool sets the MDY format
-- to ensure consistent date formatting, so we must manually override this
SET datestyle TO 'ISO, DMY';
\if :{?DEBUG}
SET client_min_messages TO debug1;
\else
SET client_min_messages TO NOTICE;
\endif
\echo Add users for testing purposes
Add users for testing purposes
SELECT * FROM public.statbus_user_create('test.super@statbus.org', 'super_user'::statbus_role_type, 'Super#123!');
         email          |  password  
------------------------+------------
 test.super@statbus.org | Super#123!
(1 row)

SELECT * FROM public.statbus_user_create('test.regular@statbus.org', 'regular_user'::statbus_role_type, 'Regular#123!');
          email           |   password   
--------------------------+--------------
 test.regular@statbus.org | Regular#123!
(1 row)

SELECT * FROM public.statbus_user_create('test.restricted@statbus.org', 'restricted_user'::statbus_role_type, 'Restricted#123!');
            email            |    password     
-----------------------------+-----------------
 test.restricted@statbus.org | Restricted#123!
(1 row)

\echo "Setting up Statbus to test enterprise grouping and primary"
"Setting up Statbus to test enterprise grouping and primary"
-- A Super User configures statbus.
CALL test.set_user_from_email('test.super@statbus.org');
\echo "User selected the Activity Category Standard"
"User selected the Activity Category Standard"
INSERT INTO settings(activity_category_standard_id,only_one_setting)
SELECT id, true FROM activity_category_standard WHERE code = 'nace_v2.1'
ON CONFLICT (only_one_setting)
DO UPDATE SET
   activity_category_standard_id =(SELECT id FROM activity_category_standard WHERE code = 'nace_v2.1')
   WHERE settings.id = EXCLUDED.id;
;
SELECT acs.code
  FROM public.settings AS s
  JOIN activity_category_standard AS acs
    ON s.activity_category_standard_id = acs.id;
   code    
-----------
 nace_v2.1
(1 row)

\echo "User uploads the sample activity categories"
"User uploads the sample activity categories"
\copy public.activity_category_available_custom(path,name,description) FROM 'samples/norway/activity_category/activity_category_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.activity_category_available;
 count 
-------
  2215
(1 row)

\echo "User uploads the sample regions"
"User uploads the sample regions"
\copy public.region_upload(path, name) FROM 'samples/norway/regions/norway-regions-2024.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.region;
 count 
-------
   417
(1 row)

\echo "User uploads the sample legal forms"
"User uploads the sample legal forms"
\copy public.legal_form_custom_only(code,name) FROM 'samples/norway/legal_form/legal_form_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.legal_form_available;
 count 
-------
    46
(1 row)

\echo "User uploads the sample sectors"
"User uploads the sample sectors"
\copy public.sector_custom_only(path,name,description) FROM 'samples/norway/sector/sector_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.sector_available;
 count 
-------
    33
(1 row)

SAVEPOINT before_loading_units;
\echo "Test deaths at the end of the year"
"Test deaths at the end of the year"
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;
 establishment_count | legal_unit_count | enterprise_count 
---------------------+------------------+------------------
                   0 |                0 |                0
(1 row)

\echo "User uploads the legal units"
"User uploads the legal units"
\copy public.import_legal_unit_era(valid_from,valid_to,tax_ident,name,birth_date,death_date,physical_address_part1,physical_postcode,physical_postplace,physical_region_code,physical_country_iso_2,postal_address_part1,postal_postcode,postal_postplace,postal_region_code,postal_country_iso_2,primary_activity_category_code,secondary_activity_category_code,sector_code,legal_form_code) FROM 'test/data/42_legal-units-deaths-end-of-year.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;
 establishment_count | legal_unit_count | enterprise_count 
---------------------+------------------+------------------
                   0 |                2 |                2
(1 row)

\echo Run worker processing to run import jobs and generate computed data
Run worker processing to run import jobs and generate computed data
CALL worker.process_tasks();
SELECT queue, state, count(*) FROM worker.tasks AS t JOIN worker.command_registry AS c ON t.command = c.command GROUP BY queue,state ORDER BY queue,state;
    queue    |   state   | count 
-------------+-----------+-------
 analytics   | completed |     7
 maintenance | pending   |     1
(2 rows)

\echo "Check statistical units"
"Check statistical units"
SELECT external_idents ->> 'tax_ident' as tax_ident, name, valid_after, valid_from, valid_to, birth_date, death_date
FROM public.statistical_unit
WHERE unit_type = 'legal_unit'
ORDER BY external_idents ->> 'tax_ident', valid_from;
 tax_ident |       name        | valid_after | valid_from |  valid_to  | birth_date | death_date 
-----------+-------------------+-------------+------------+------------+------------+------------
 823573673 | Kranløft Vestland | 2009-12-31  | 2010-01-01 | 2012-12-30 | 2010-01-01 | 
 823573673 | Kranløft Vestland | 2012-12-30  | 2012-12-31 | 2012-12-31 | 2010-01-01 | 2012-12-31
 921835809 | Kranløft Østland  | 2009-12-31  | 2010-01-01 | 2011-12-30 | 2010-01-01 | 
 921835809 | Kranløft Østland  | 2011-12-30  | 2011-12-31 | 2011-12-31 | 2010-01-01 | 2011-12-31
(4 rows)

\echo "Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
"Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year'
AND year < 2013
AND unit_type = 'legal_unit';
 resolution | year | month | unit_type  | count | births | deaths 
------------+------+-------+------------+-------+--------+--------
 year       | 2010 |       | legal_unit |     2 |      2 |      0
 year       | 2011 |       | legal_unit |     1 |      0 |      1
 year       | 2012 |       | legal_unit |     0 |      0 |      1
(3 rows)

\echo "Check statistical unit history by year-month - deaths should be 1 for year-month 2011-12 and 2012-12"
"Check statistical unit history by year-month - deaths should be 1 for year-month 2011-12 and 2012-12"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year-month'
AND year < 2013
AND unit_type = 'legal_unit';
 resolution | year | month | unit_type  | count | births | deaths 
------------+------+-------+------------+-------+--------+--------
 year-month | 2010 |     1 | legal_unit |     2 |      2 |      0
 year-month | 2010 |     2 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     3 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     4 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     5 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     6 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     7 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     8 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     9 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    10 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    11 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    12 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     1 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     2 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     3 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     4 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     5 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     6 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     7 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     8 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     9 | legal_unit |     2 |      0 |      0
 year-month | 2011 |    10 | legal_unit |     2 |      0 |      0
 year-month | 2011 |    11 | legal_unit |     2 |      0 |      0
 year-month | 2011 |    12 | legal_unit |     1 |      0 |      1
 year-month | 2012 |     1 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     2 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     3 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     4 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     5 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     6 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     7 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     8 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     9 | legal_unit |     1 |      0 |      0
 year-month | 2012 |    10 | legal_unit |     1 |      0 |      0
 year-month | 2012 |    11 | legal_unit |     1 |      0 |      0
 year-month | 2012 |    12 | legal_unit |     0 |      0 |      1
(36 rows)

\x
ROLLBACK TO before_loading_units;
\echo "Test deaths at the end of the first month"
"Test deaths at the end of the first month"
\x
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;
 establishment_count | legal_unit_count | enterprise_count 
---------------------+------------------+------------------
                   0 |                0 |                0
(1 row)

\echo "User uploads the legal units"
"User uploads the legal units"
\copy public.import_legal_unit_era(valid_from,valid_to,tax_ident,name,birth_date,death_date,physical_address_part1,physical_postcode,physical_postplace,physical_region_code,physical_country_iso_2,postal_address_part1,postal_postcode,postal_postplace,postal_region_code,postal_country_iso_2,primary_activity_category_code,secondary_activity_category_code,sector_code,legal_form_code) FROM 'test/data/42_legal-units-deaths-end-of-first-month.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;
 establishment_count | legal_unit_count | enterprise_count 
---------------------+------------------+------------------
                   0 |                2 |                2
(1 row)

\echo Run worker processing to run import jobs and generate computed data
Run worker processing to run import jobs and generate computed data
CALL worker.process_tasks();
SELECT queue, state, count(*) FROM worker.tasks AS t JOIN worker.command_registry AS c ON t.command = c.command GROUP BY queue,state ORDER BY queue,state;
    queue    |   state   | count 
-------------+-----------+-------
 analytics   | completed |     7
 maintenance | pending   |     1
(2 rows)

\echo "Check statistical units"
"Check statistical units"
SELECT external_idents ->> 'tax_ident' as tax_ident, name, valid_after, valid_from, valid_to, birth_date, death_date
FROM public.statistical_unit
WHERE unit_type = 'legal_unit'
ORDER BY external_idents ->> 'tax_ident', valid_from;
 tax_ident |       name        | valid_after | valid_from |  valid_to  | birth_date | death_date 
-----------+-------------------+-------------+------------+------------+------------+------------
 823573673 | Kranløft Vestland | 2009-12-31  | 2010-01-01 | 2012-12-30 | 2010-01-01 | 
 823573673 | Kranløft Vestland | 2012-12-30  | 2012-12-31 | 2012-12-31 | 2010-01-01 | 2012-12-31
 921835809 | Kranløft Østland  | 2009-12-31  | 2010-01-01 | 2011-01-31 | 2010-01-01 | 2011-01-31
(3 rows)

\echo "Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
"Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year'
AND year < 2013
AND unit_type = 'legal_unit';
 resolution | year | month | unit_type  | count | births | deaths 
------------+------+-------+------------+-------+--------+--------
 year       | 2010 |       | legal_unit |     2 |      2 |      0
 year       | 2011 |       | legal_unit |     1 |      0 |      1
 year       | 2012 |       | legal_unit |     0 |      0 |      1
(3 rows)

\echo "Check statistical unit history by year-month - deaths should be 1 for year-month 2011-1 and 2012-12"
"Check statistical unit history by year-month - deaths should be 1 for year-month 2011-1 and 2012-12"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year-month'
AND year < 2013
AND unit_type = 'legal_unit';
 resolution | year | month | unit_type  | count | births | deaths 
------------+------+-------+------------+-------+--------+--------
 year-month | 2010 |     1 | legal_unit |     2 |      2 |      0
 year-month | 2010 |     2 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     3 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     4 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     5 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     6 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     7 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     8 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     9 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    10 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    11 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    12 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     1 | legal_unit |     1 |      0 |      1
 year-month | 2011 |     2 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     3 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     4 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     5 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     6 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     7 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     8 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     9 | legal_unit |     1 |      0 |      0
 year-month | 2011 |    10 | legal_unit |     1 |      0 |      0
 year-month | 2011 |    11 | legal_unit |     1 |      0 |      0
 year-month | 2011 |    12 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     1 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     2 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     3 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     4 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     5 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     6 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     7 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     8 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     9 | legal_unit |     1 |      0 |      0
 year-month | 2012 |    10 | legal_unit |     1 |      0 |      0
 year-month | 2012 |    11 | legal_unit |     1 |      0 |      0
 year-month | 2012 |    12 | legal_unit |     0 |      0 |      1
(36 rows)

\x
ROLLBACK TO before_loading_units;
\echo "Test deaths in the start of a month"
"Test deaths in the start of a month"
\x
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;
 establishment_count | legal_unit_count | enterprise_count 
---------------------+------------------+------------------
                   0 |                0 |                0
(1 row)

\echo "User uploads the legal units"
"User uploads the legal units"
\copy public.import_legal_unit_era(valid_from,valid_to,tax_ident,name,birth_date,death_date,physical_address_part1,physical_postcode,physical_postplace,physical_region_code,physical_country_iso_2,postal_address_part1,postal_postcode,postal_postplace,postal_region_code,postal_country_iso_2,primary_activity_category_code,secondary_activity_category_code,sector_code,legal_form_code) FROM 'test/data/42_legal-units-deaths-start-of-month.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;
 establishment_count | legal_unit_count | enterprise_count 
---------------------+------------------+------------------
                   0 |                2 |                2
(1 row)

\echo Run worker processing to run import jobs and generate computed data
Run worker processing to run import jobs and generate computed data
CALL worker.process_tasks();
SELECT queue, state, count(*) FROM worker.tasks AS t JOIN worker.command_registry AS c ON t.command = c.command GROUP BY queue,state ORDER BY queue,state;
    queue    |   state   | count 
-------------+-----------+-------
 analytics   | completed |     7
 maintenance | pending   |     1
(2 rows)

\echo "Check statistical units"
"Check statistical units"
SELECT external_idents ->> 'tax_ident' as tax_ident, name, valid_after, valid_from, valid_to, birth_date, death_date
FROM public.statistical_unit
WHERE unit_type = 'legal_unit'
ORDER BY external_idents ->> 'tax_ident', valid_from;
 tax_ident |       name        | valid_after | valid_from |  valid_to  | birth_date | death_date 
-----------+-------------------+-------------+------------+------------+------------+------------
 823573673 | Kranløft Vestland | 2009-12-31  | 2010-01-01 | 2011-12-31 | 2010-01-01 | 
 823573673 | Kranløft Vestland | 2011-12-31  | 2012-01-01 | 2012-01-01 | 2010-01-01 | 2012-01-01
 921835809 | Kranløft Østland  | 2009-12-31  | 2010-01-01 | 2011-01-01 | 2010-01-01 | 2011-01-01
(3 rows)

\echo "Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
"Check statistical unit history by year - deaths should be 1 for year 2011 and 2012"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year'
AND year < 2013
AND unit_type = 'legal_unit';
 resolution | year | month | unit_type  | count | births | deaths 
------------+------+-------+------------+-------+--------+--------
 year       | 2010 |       | legal_unit |     2 |      2 |      0
 year       | 2011 |       | legal_unit |     1 |      0 |      1
 year       | 2012 |       | legal_unit |     0 |      0 |      1
(3 rows)

\echo "Check statistical unit history by year-month - deaths should be 1 for year-month 2011-1 and 2012-1"
"Check statistical unit history by year-month - deaths should be 1 for year-month 2011-1 and 2012-1"
SELECT resolution, year, month, unit_type, count, births, deaths
FROM public.statistical_history
WHERE resolution = 'year-month'
AND year < 2013
AND unit_type = 'legal_unit';
 resolution | year | month | unit_type  | count | births | deaths 
------------+------+-------+------------+-------+--------+--------
 year-month | 2010 |     1 | legal_unit |     2 |      2 |      0
 year-month | 2010 |     2 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     3 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     4 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     5 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     6 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     7 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     8 | legal_unit |     2 |      0 |      0
 year-month | 2010 |     9 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    10 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    11 | legal_unit |     2 |      0 |      0
 year-month | 2010 |    12 | legal_unit |     2 |      0 |      0
 year-month | 2011 |     1 | legal_unit |     1 |      0 |      1
 year-month | 2011 |     2 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     3 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     4 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     5 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     6 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     7 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     8 | legal_unit |     1 |      0 |      0
 year-month | 2011 |     9 | legal_unit |     1 |      0 |      0
 year-month | 2011 |    10 | legal_unit |     1 |      0 |      0
 year-month | 2011 |    11 | legal_unit |     1 |      0 |      0
 year-month | 2011 |    12 | legal_unit |     1 |      0 |      0
 year-month | 2012 |     1 | legal_unit |     0 |      0 |      1
(25 rows)

\x
ROLLBACK;
